File: tests/output/python/tests/client/search_test.py
from re import compile

from algoliasearch.http.hosts import Host, HostsCollection
from algoliasearch.http.transporter import EchoTransporter
from algoliasearch.search.client import SearchClient
from algoliasearch.search.config import SearchConfig


class TestSearchClient:
    _client: SearchClient

    def create_client(self) -> SearchClient:
        _config = SearchConfig("appId", "apiKey")
        self._client = SearchClient.create_with_config(
            config=_config, transporter=EchoTransporter(_config)
        )

    async def test_api_0(self):
        """
        calls api with correct read host
        """

        _config = SearchConfig("test-app-id", "test-api-key")
        self._client = SearchClient.create_with_config(
            config=_config, transporter=EchoTransporter(_config)
        )
        _req = await self._client.custom_get_with_http_info(
            path="/test",
        )
        assert _req.host == "test-app-id-dsn.algolia.net"

    async def test_api_1(self):
        """
        calls api with correct write host
        """

        _config = SearchConfig("test-app-id", "test-api-key")
        self._client = SearchClient.create_with_config(
            config=_config, transporter=EchoTransporter(_config)
        )
        _req = await self._client.custom_post_with_http_info(
            path="/test",
        )
        assert _req.host == "test-app-id.algolia.net"

    async def test_api_2(self):
        """
        tests the retry strategy
        """

        _config = SearchConfig("test-app-id", "test-api-key")
        _config.hosts = HostsCollection(
            [
                Host(url="localhost", scheme="http", port=6677),
                Host(url="localhost", scheme="http", port=6678),
            ]
        )
        self._client = SearchClient.create_with_config(config=_config)
        _req = await self._client.custom_get(
            path="/test",
        )
        assert _req == """{"message":"ok test server response"}"""

    async def test_common_api_0(self):
        """
        calls api with correct user agent
        """
        self.create_client()

        _req = await self._client.custom_post_with_http_info(
            path="/test",
        )
        regex_user_agent = compile(
            "^Algolia for Python \\(\\d+\\.\\d+\\.\\d+(-?.*)?\\)(; [a-zA-Z. ]+ (\\(\\d+((\\.\\d+)?\\.\\d+)?(-?.*)?\\))?)*(; Search (\\(\\d+\\.\\d+\\.\\d+(-?.*)?\\)))(; [a-zA-Z. ]+ (\\(\\d+((\\.\\d+)?\\.\\d+)?(-?.*)?\\))?)*$"
        )
        assert regex_user_agent.match(_req.headers.get("user-agent")) is not None

    async def test_common_api_1(self):
        """
        calls api with default read timeouts
        """
        self.create_client()

        _req = await self._client.custom_get_with_http_info(
            path="/test",
        )
        assert _req.timeouts.get("connect") == 2000
        assert _req.timeouts.get("response") == 5000

    async def test_common_api_2(self):
        """
        calls api with default write timeouts
        """
        self.create_client()

        _req = await self._client.custom_post_with_http_info(
            path="/test",
        )
        assert _req.timeouts.get("connect") == 2000
        assert _req.timeouts.get("response") == 30000

    async def test_parameters_0(self):
        """
        client throws with invalid parameters
        """

        try:
            _config = SearchConfig("", "")
            self._client = SearchClient.create_with_config(
                config=_config, transporter=EchoTransporter(_config)
            )
            assert False
        except (ValueError, Exception) as e:
            assert str(e) == "`app_id` is missing."
        try:
            _config = SearchConfig("", "my-api-key")
            self._client = SearchClient.create_with_config(
                config=_config, transporter=EchoTransporter(_config)
            )
            assert False
        except (ValueError, Exception) as e:
            assert str(e) == "`app_id` is missing."
        try:
            _config = SearchConfig("my-app-id", "")
            self._client = SearchClient.create_with_config(
                config=_config, transporter=EchoTransporter(_config)
            )
            assert False
        except (ValueError, Exception) as e:
            assert str(e) == "`api_key` is missing."

    async def test_parameters_1(self):
        """
        &#x60;addApiKey&#x60; throws with invalid parameters
        """
        self.create_client()

        try:
            await self._client.add_api_key_with_http_info(
                api_key=None,
            )
            assert False
        except (ValueError, Exception) as e:
            assert (
                str(e) == "Parameter `api_key` is required when calling `add_api_key`."
            )

    async def test_parameters_2(self):
        """
        &#x60;addOrUpdateObject&#x60; throws with invalid parameters
        """
        self.create_client()

        try:
            await self._client.add_or_update_object_with_http_info(
                index_name=None,
                object_id="my-object-id",
                body={},
            )
            assert False
        except (ValueError, Exception) as e:
            assert (
                str(e)
                == "Parameter `index_name` is required when calling `add_or_update_object`."
            )
        try:
            await self._client.add_or_update_object_with_http_info(
                index_name="my-index-name",
                object_id=None,
                body={},
            )
            assert False
        except (ValueError, Exception) as e:
            assert (
                str(e)
                == "Parameter `object_id` is required when calling `add_or_update_object`."
            )
        try:
            await self._client.add_or_update_object_with_http_info(
                index_name="my-index-name",
                object_id="my-object-id",
                body=None,
            )
            assert False
        except (ValueError, Exception) as e:
            assert (
                str(e)
                == "Parameter `body` is required when calling `add_or_update_object`."
            )

File: tests/output/scala/src/test/scala/algoliasearch/client/SearchTest.scala
package algoliasearch.client

import algoliasearch.api.SearchClient
import algoliasearch.config.*
import algoliasearch.{EchoInterceptor, assertError}
import algoliasearch.search.*
import algoliasearch.exception.*
import org.json4s.*
import org.json4s.native.JsonParser.*
import org.scalatest.funsuite.AnyFunSuite

import scala.concurrent.duration.Duration
import scala.concurrent.{Await, ExecutionContextExecutor}

class SearchTest extends AnyFunSuite {
  implicit val ec: ExecutionContextExecutor = scala.concurrent.ExecutionContext.global
  implicit val formats: Formats = org.json4s.DefaultFormats

  def testClient(appId: String = "appId", apiKey: String = "apiKey"): (SearchClient, EchoInterceptor) = {
    val echo = EchoInterceptor()
    (
      SearchClient(
        appId = appId,
        apiKey = apiKey,
        clientOptions = ClientOptions
          .builder()
          .withRequesterConfig(requester => requester.withInterceptor(echo))
          .build()
      ),
      echo
    )
  }

  test("calls api with correct read host") {

    val (client, echo) = testClient(appId = "test-app-id", apiKey = "test-api-key")

    Await.ready(
      client.customGet[Any](
        path = "/test"
      ),
      Duration.Inf
    )
    assert(echo.lastResponse.get.host == "test-app-id-dsn.algolia.net")
  }

  test("calls api with correct write host") {

    val (client, echo) = testClient(appId = "test-app-id", apiKey = "test-api-key")

    Await.ready(
      client.customPost[Any](
        path = "/test"
      ),
      Duration.Inf
    )
    assert(echo.lastResponse.get.host == "test-app-id.algolia.net")
  }

  test("calls api with correct user agent") {
    val (client, echo) = testClient()

    Await.ready(
      client.customPost[Any](
        path = "/test"
      ),
      Duration.Inf
    )
    val regexp =
      """^Algolia for Scala \(\d+\.\d+\.\d+(-?.*)?\)(; [a-zA-Z. ]+ (\(\d+((\.\d+)?\.\d+)?(-?.*)?\))?)*(; Search (\(\d+\.\d+\.\d+(-?.*)?\)))(; [a-zA-Z. ]+ (\(\d+((\.\d+)?\.\d+)?(-?.*)?\))?)*$""".r
    val header = echo.lastResponse.get.headers("user-agent")
    assert(header.matches(regexp.regex), s"Expected $header to match the following regex: ${regexp.regex}")
  }

  test("calls api with default read timeouts") {
    val (client, echo) = testClient()

    Await.ready(
      client.customGet[Any](
        path = "/test"
      ),
      Duration.Inf
    )
    assert(echo.lastResponse.get.connectTimeout == 2000)
    assert(echo.lastResponse.get.responseTimeout == 5000)
  }

  test("calls api with default write timeouts") {
    val (client, echo) = testClient()

    Await.ready(
      client.customPost[Any](
        path = "/test"
      ),
      Duration.Inf
    )
    assert(echo.lastResponse.get.connectTimeout == 2000)
    assert(echo.lastResponse.get.responseTimeout == 30000)
  }

  test("client throws with invalid parameters") {

    assertError("`appId` is missing.") {
      val (client, echo) = testClient(appId = "", apiKey = "")
    }

    assertError("`appId` is missing.") {
      val (client, echo) = testClient(appId = "", apiKey = "my-api-key")
    }

    assertError("`apiKey` is missing.") {
      val (client, echo) = testClient(appId = "my-app-id", apiKey = "")
    }
  }

  test("`addApiKey` throws with invalid parameters") {
    val (client, echo) = testClient()

    assertError("Parameter `apiKey` is required when calling `addApiKey`.") {
      Await.ready(
        client.addApiKey(
          apiKey = null
        ),
        Duration.Inf
      )
    }
  }

  test("`addOrUpdateObject` throws with invalid parameters") {
    val (client, echo) = testClient()

    assertError("Parameter `indexName` is required when calling `addOrUpdateObject`.") {
      Await.ready(
        client.addOrUpdateObject(
          indexName = null,
          objectID = "my-object-id",
          body = JObject(List())
        ),
        Duration.Inf
      )
    }

    assertError("Parameter `objectID` is required when calling `addOrUpdateObject`.") {
      Await.ready(
        client.addOrUpdateObject(
          indexName = "my-index-name",
          objectID = null,
          body = JObject(List())
        ),
        Duration.Inf
      )
    }

    assertError("Parameter `body` is required when calling `addOrUpdateObject`.") {
      Await.ready(
        client.addOrUpdateObject(
          indexName = "my-index-name",
          objectID = "my-object-id",
          body = null
        ),
        Duration.Inf
      )
    }
  }
}
